### PR TITLE
[Barracuda CloudGen] Add initial Barracuda CloudGen Firewall integration

### LINKED ISSUE #3773: Barracuda Cloudgen Firewall
### Description
Barracuda CloudGen Firewall provides multiple layers of detection including advanced threat signatures, behavioral and heuristic analysis, static code analysis, and finally comprehensive sandboxing, to provide accurate detection and in-depth protection against ransomware, malware, and other advanced cyber-attacks.

### Architecture
CloudGen ships with Filebeat which processes data on the CloudGen side into structured JSON data. They then output to Logstash via the Lumberjack protocol (see steps [here](https://campus.barracuda.com/product/cloudgenfirewall/doc/96025108/how-to-enable-filebeat-stream-to-a-logstash-pipeline/)).  With this integration, we're proposing to add a lumberjack input to Filebeat, which could receive the structured events from CloudGen. We then build a Fleet integration with ingest pipelines and dashboards. 

An alternative approach is their syslog output, but the data doesn't conform to the syslog RFC's and may be difficult to parse. There would also be more maintenance involved as their syslog parsing often changes with major releases. 

# Integration release checklist

This checklist is intended for integrations maintainers to ensure consistency
when creating or updating a Package, Module or Dataset for an Integration.

### All changes

- [ ] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md)
- [ ] Supported versions of the monitoring target are documented
- [ ] Supported operating systems are documented (if applicable)
- [ ] Integration or [System tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md) exist
- [ ] Documentation exists
- [ ] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
- [ ] At least a manual test with ES / Kibana / Agent has been performed.
- [ ] Required Kibana version set to:

### New Package

- [ ] Screenshot of the "Add Integration" page on Fleet added

### Dashboards changes

- [ ] Dashboards exists
- [ ] Screenshots added or updated
- [ ] Datastream filters added to visualizations

### Log dataset changes

- [ ] [Pipeline tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/pipeline_testing.md) exist (if applicable)
- [ ] Generated output for at least 1 log file exists
- [ ] Sample event (`sample_event.json`) exists

### PR DESCRIPTION
**Pull request by legoguy1000**
## What does this PR do?

Add initial Barracuda CloudGen Firewall integration for receiving Firewall Insight logs as described at https://campus.barracuda.com/product/cloudgenfirewall/doc/96025953/how-to-enable-filebeat-stream-to-a-logstash-pipeline. Elastic Agent starts a server to receive data sent over the Lumberjack protocol by CloudGen firewall. (This is the same protocol used between Beats and Logstash.)

## Checklist

- [X] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [X] I have verified that all data streams collect metrics or logs.
- [X] I have added an entry to my package's `changelog.yml` file.
- [X] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Requires elastic/beats#32175
- Closes #3773

## Screenshots

Screen Shot 2022-09-21 at 09 56 14

cloudgen-config-page

### CODE REVIEW

**Comment by legoguy1000** [comment]
Tried using the lumberjack input today with the agent and I guess it needs to be added to this list first https://github.com/elastic/elastic-agent/blob/main/internal/spec/filebeat.yml#L24???

**Comment by andrewkroh** [comment]
Good call. That spec needs updated.

**Comment by legoguy1000** [comment]
@andrewkroh Using the 8.5.0-SNAPSHOT image, the lumberjack input is working. I am running into 2 issues though.
1.  I am currently using a filebeat container to simulate the lumberjack source but there appears to be a race condition with Filebeat and the Agent setup/policy... Unlike with elastic/stream which waits for the SIGHUP, Filebeat doesn't wait and I can't get events to generate unless I manually restart the Filebeat container and "republish" the events.
2. I'm getting some publishing errors
```
[elastic_agent.filebeat][warn] Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Date(2022, time.August, 25, 23, 56, 33, 95198956, time.UTC), Meta:{"input_id":"lumberjack-a08f22ef-5d2d-4ffc-8fb8-3d5cb2aa7f91","raw_index":"logs-barracuda_cloudgen_firewall.log-ep","stream_id":"lumberjack-barracuda_cloudgen_firewall.log-a08f22ef-5d2d-4ffc-8fb8-3d5cb2aa7f91"}, Fields:{"agent":{"ephemeral_id":"6ddcb0a4-6f36-4b37-8395-aaa73a6ba832","id":"4d7d11a6-0d0f-4694-8c64-787b073cfc45","name":"docker-fleet-agent","type":"filebeat","version":"8.5.0"},"data_stream":{"dataset":"barracuda_cloudgen_firewall.log","namespace":"ep","type":"logs"},"ecs":{"version":"8.0.0"},"elastic_agent":{"id":"4d7d11a6-0d0f-4694-8c64-787b073cfc45","snapshot":true,"version":"8.5.0"},"event":{"dataset":"barracuda_cloudgen_firewall.log"},"input":{"type":"lumberjack"},"lumberjack":{"@metadata":{"beat":"filebeat","type":"_doc","version":"8.3.3"},"@timestamp":"2022-08-25T23:56:32.088Z","agent":{"ephemeral_id":"47e01448-f2e5-4f53-b71f-ee50576d7028","id":"17ee087a-c1dd-4d73-90c6-c8f9c81c1e02","name":"8761db0f577a","type":"filebeat","version":"8.3.3"},"ecs":{"version":"8.0.0"},"host":{"name":"8761db0f577a"},"input":{"type":"filestream"},"log":{"file":{"path":"/sample_logs/web.log"},"offset":467},"message":"{\"timestamp\":1526377804000,\"traffic_type\":0,\"action\":0,\"source_ip\":\"192.168.42.105\",\"source_port\":\"50159\",\"destination_ip\":\"89.160.20.114\",\"destination_port\":\"443\",\"method\":\"GET\",\"status_code\":\"200\",\"user_agent\":\"mozilla/5.0 (windows nt 6.1) applewebkit/537.36 (khtml, like gecko) chrome/66.0.3359.139 safari/537.36\",\"content_type\":\"\",\"name\":\"https://clientservices.googleapis.com/chrome-variations/seed?osname=win\u0026channel=stable\u0026milestone=66\",\"size\":0,\"domain\":\"clientservices.googleapis.com\",\"category\":[],\"user\":\"192.168.42.105\",\"user_type\":0,\"fw_rule\":\"LAN-2-INTERNET\",\"app_rule\":\"\u003cApp\u003e:\u003cpass-no-match\u003e\"}","product":"ngfw","sn":"4f94abdf7a8c465fa2cd76f680ecafd1","type":"ngfw-wf"},"source":{"address":"192.168.16.5:53796"},"tags":["barracuda_cloudgen_firewall-log","forwarded"]}, Private:(*lumberjack.batchACKTracker)(0xc000011008), TimeSeries:false}, Flags:0x1, Cache:publisher.EventCache{m:mapstr.M(nil)}} (status=400): {"type":"mapper_parsing_exception","reason":"failed to parse field [@timestamp] of type [date] in document with id '9Qxv14IBNgppDiFepmQJ'. Preview of field's value: '+50338-12-27T18:26:40.000Z'","caused_by":{"type":"illegal_argument_exception","reason":"failed to parse date field [+50338-12-27T18:26:40.000Z] with format [strict_date_optional_time||epoch_millis]","caused_by":{"type":"date_time_parse_exception","reason":"Failed to parse with all enclosed parsers"}}}, dropping event!
```

**Comment by andrewkroh** [comment]
We can add in a lumberjack output to elastic/stream to ease testing. I've opened https://github.com/elastic/stream/issues/39 to track this.

> I'm getting some publishing errors

The event that is logged doesn't show that odd year value so I wonder if it's something happening on the ingest node pipeline side that's causing the data corruption.

**Comment by legoguy1000** [comment]
I think I may have found it.  The Web data uses Epoch MS, not Seconds. Also I may have a way to wait for the agent to be ready.

**Comment by legoguy1000** [comment]
Error 1 is resolved.  My plan for the race condition seems to kinda work but I think there is a bigger problem as I'm getting multiple (5+) log records of the lumberjack input starting and stopping and I think thats causing some of the issues.

image

**Comment by epixa** [comment]
Can you add this integration to the CODEOWNERS file in this PR?

**Comment by andrewkroh** [comment]
I added the Lumberjack output to elastic/stream in https://github.com/elastic/stream/pull/41. This is the config that I was using. I got blocked by an issue in 8.5.0-SNAPSHOT today where the API keys for Filebeat were invalid, so I couldn't test E2E. But I did confirm that stream was connected and communicating to Filebeat.

```yaml
version: '2.3'
services:
  barracuda-cloudgen-lumberjack:
    image: docker.elastic.co/observability/stream:local # Requires next release.
    volumes:
      - ./sample_logs:/sample_logs:ro
    environment:
      - STREAM_PROTOCOL=lumberjack
      - STREAM_LUMBERJACK_PARSE_JSON=true
      - STREAM_ADDR=tcp://elastic-agent:5044
      - STREAM_DELAY=5s
      - STREAM_START_SIGNAL=SIGHUP
    command: log /sample_logs/*.log
```

**Comment by legoguy1000** [comment]
Using ur example, the JSON in the sample logs file was sent in the `message` block to the agent? Is there a method to send additional fields? The CloudGen is sending additional fields separate from the message field, https://campus.barracuda.com/product/cloudgenfirewall/doc/96025108/how-to-enable-filebeat-stream-to-a-logstash-pipeline/, also shown in my POC using Filebeat.

**Comment by andrewkroh** [comment]
Using my example stream config, each JSON log line will be sent as a structured event. On receiving side, the lumberjack input will produce an event that contains `lumberjack.*` whose contents are everything from the JSON log line (as an object).

**Comment by andrewkroh** [comment]
The new container with lumberjack output is ready: `docker.elastic.co/observability/stream:v0.8.0`

**Comment by legoguy1000** [comment]
@andrewkroh I think ready to try tests. I sent a message in slack with an error that the stream container seems to have
```
{"level":"info","ts":"2022-09-19T18:01:29.890Z","caller":"command/root.go:138","msg":"Delaying connection.","delay":5}
{"level":"debug","ts":"2022-09-19T18:01:34.891Z","caller":"output/util.go:28","msg":"Connecting...","address":"tcp://elastic-agent:5044"}
{"level":"info","ts":"2022-09-19T18:01:34.894Z","caller":"output/util.go:41","msg":"Connected","address":"tcp://elastic-agent:5044"}
{"level":"debug","ts":"2022-09-19T18:01:34.895Z","caller":"command/log.go:81","msg":"Sending log line.","address":"tcp://elastic-agent:5044","log":"/sample_logs/firewall.log","line_number":1}
Error: EOF
```
It appears to hang on the `sending log line....`

**Comment by legoguy1000** [comment]
@andrewkroh Here appears to be the panic from Filebeat
```
{"log.level":"error","@timestamp":"2022-09-20T02:21:21.793Z","log.origin":{"file.name":"process/stdlogger.go","file.line":54},"message":"filebeat stderr: \"panic: close of closed channel\\n\\ngoroutine 130 [running]:\\ngithub.com/elastic/beats/v7/filebeat/beater.(*Filebeat).Stop(0xc000a09710)\\n\\t/go/src/github.com/elastic/beats/filebeat/beater/filebeat.go:426 +0x46\\ngithub.com/elastic/beats/v7/libbeat/cmd/instance.(*Beat).launch.func5()\\n\\t/go/src/github.com/elastic/beats/libbeat/cmd/instance/beat.go:461 +0x68\\nsync.(*Once).doSlow(0xc00007c5d0, 0xc000080480)\\n\\t/usr/local/go/src/sync/once.go:68 +0x138\\nsync.(*Once).Do(0xc00007c5d0, 0xc000080480)\\n\\t/usr/local/go/src/sync/once.go:59 +0x45\\ngithub.com/elastic/elastic-agent-libs/service.HandleSignals.func1()\\n\\t/go/pkg/mod/github.com/elastic/elastic-agent-libs@v0.2.11/service/service.go:\"","agent.console.name":"filebeat","agent.console.type":"stderr","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2022-09-20T02:21:21.795Z","log.origin":{"file.name":"process/stdlogger.go","file.line":54},"message":"filebeat stderr: \"60 +0x20c\\ncreated by github.com/elastic/elastic-agent-libs/service.HandleSignals\\n\"","agent.console.name":"filebeat","agent.console.type":"stderr","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2022-09-20T02:21:21.795Z","log.origin":{"file.name":"process/stdlogger.go","file.line":54},"message":"filebeat stderr: \"\\t/go/pkg/mod/github.com/elastic/elastic-agent-libs@v0.2.11/service/service.go:49 +0x25c\\n\"","agent.console.name":"filebeat","agent.console.type":"stderr","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-20T02:21:21.804Z","log.origin":{"file.name":"log/reporter.go","file.line":40},"message":"2022-09-20T02:21:21Z - message: Application: filebeat--8.5.0-SNAPSHOT[e4df6077-3e31-4c83-b11d-8d2d23282840]: State changed to STOPPED: Stopped - type: 'STATE' - sub_type: 'STOPPED'","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-20T02:21:22.364Z","log.origin":{"file.name":"fleet/fleet_gateway.go","file.line":386},"message":"Fleet gateway is stopping","ecs.version":"1.6.0"}
```

**Comment by andrewkroh** [comment]
After reviewing those panic stack traces I think they are all related to https://github.com/elastic/beats/issues/32608. None of them appear to be related to the the `lumberjack` package. I suspect reason the `elastic/stream` tester is blocked because Filebeat is not ACKing due to https://github.com/elastic/fleet-server/issues/1877.

File: packages/barracuda_cloudgen_firewall/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      ignore_missing: true
+  - rename:
+      field: tls.client.subject
+      target_field: labels.origin_client_subject
+      ignore_missing: true
+  - remove:
+      field:
+        - source
+        - tls
+      ignore_missing: true
```
Comment (legoguy1000) [inline packages/barracuda_cloudgen_firewall/data_stream/log/elasticsearch/ingest_pipeline/default.yml:21]: Separate from this PR would it make sense to update the lumberjack input to use different fields than source and tls in the actual input code? It seems misleading by default to have those fields populated when its only relevant to the transport of receiving the logs, not really whats in the log its self? Just a thought.

Comment (andrewkroh) [reply packages/barracuda_cloudgen_firewall/data_stream/log/elasticsearch/ingest_pipeline/default.yml:21]: I was thinking to propose some official ECS fields for this data like perhaps `log.source.ip`, `log.source.port`, and `log.source.x509.*`.

**Review by andrewkroh — APPROVED** [review]
I think this is good to go for a technical preview. Things to follow up on:

- Propose standardized fields for log origin metadata.
- Test with a real CloudGen firewall.
- Document TLS requirements if any (I think the CloudGen UI requires TLS enabled on the server.)

**Comment by andrewkroh** [comment]
Great work @legoguy1000 getting to the finishing line. Thank you!

**Pull request merged**